### PR TITLE
Blocks only drop pickups when using correct tool

### LINKED ISSUE #4475: Players get item drops without using pickaxe
Client version: 1.12.2
Server OS: Linux
Cuberite Commit id: https://github.com/cuberite/cuberite/commit/ddc92ec0345f73f9e834fffc346a5bdb6f3fa3bc

### Expected behavior
Player must use a pickaxe to receive drops from blocks like stone and ores.

### Actual behavior
Item drops are created when a player breaks a block using their hand.

### Steps to reproduce the behavior
Mine a stone block without holding a pickaxe.

Regression introduced in https://github.com/cuberite/cuberite/pull/4417
Seems like CanHarvestBlock is not called anymore?

### PR DESCRIPTION
**Pull request by maxluchterhand1**
Fixes #4475

### CODE REVIEW

File: src/World.cpp [hunk 1]
```diff
@@ -2201,7 +2201,8 @@ bool cWorld::DigBlock(Vector3i a_BlockPos)
 bool cWorld::DropBlockAsPickups(Vector3i a_BlockPos, const cEntity * a_Digger, const cItem * a_Tool)
 {
 	auto pickups = PickupsFromBlock(a_BlockPos, a_Digger, a_Tool);
-	if (!DigBlock(a_BlockPos))
+	bool CanHarvestBlock {a_Tool->GetHandler()->CanHarvestBlock(GetBlock(a_BlockPos))};
```
Comment (peterbell10) [inline src/World.cpp]: Crash if `a_Tool` is `nullptr`.

File: src/World.cpp [hunk 2]
```diff
@@ -2201,7 +2201,8 @@ bool cWorld::DigBlock(Vector3i a_BlockPos)
...
+	if (!CanHarvestBlock || !DigBlock(a_BlockPos))
 	{
 		return false;
```
Comment (peterbell10) [inline src/World.cpp]: Docs say this function only returns `false` if the chunk is not loaded.

File: src/World.cpp (hunk 1 above)
Comment (maxluchterhand1) [reply src/World.cpp]: Fixed

File: src/World.cpp (hunk 2 above)
Comment (maxluchterhand1) [reply src/World.cpp]: Check now happens in cChunk::PickupsFromBlock, returning empty list if tool is not valid

File: src/Chunk.cpp [hunk 3]
```diff
@@ -974,7 +974,21 @@ cItems cChunk::PickupsFromBlock(Vector3i a_RelPos, const cEntity * a_Digger, con
...
+	cItems pickups (0);
+	bool canHarvestBlock;
+	if (a_Tool)
+	{
+		auto toolHandler = a_Tool->GetHandler();
+		canHarvestBlock = toolHandler->CanHarvestBlock(blockType);
+	}
+	else
+	{
+		canHarvestBlock = false;
```
Comment (peterbell10) [inline src/Chunk.cpp]: Some blocks can be harvested with an empty hand. The empty handler should cover this:
```cpp
auto toolHandler = a_Tool ? a_Tool->GetHandler() : cItemHandler::GetItemHandler(E_ITEM_EMPTY);
```

Comment (maxluchterhand1) [reply src/Chunk.cpp]: Accounted for that

File: .gitignore [hunk 4]
```diff
@@ -104,6 +105,7 @@ ReleaseProfile/
 *.dir/
 CPackConfig.cmake
 CPackSourceConfig.cmake
+cmake-build-debug
```
Comment (peterbell10) [inline .gitignore]: For future reference, you can use `.git/info/exclude` to add to have git ignore files that are only relevant to you.

Comment (maxluchterhand1) [reply .gitignore]: Alright, thanks for the tip

**Pull request merged**